ST Map Gestures implementation

MikeMatrix opened the pull request:
Basic implementation of ST Map Gestures.

Thanks to @Dslyecxi for the base code.

Still WIP

SzwedzikPL commented:
:+1:

bux578 commented on addons/map_gestures/functions/fnc_assignClientIDOnServer.sqf:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Author: MikeMatrix
```
Shouldn't dslyecxi been mentioned here aswell ...

MikeMatrix replied:
True, I forgot to adjust the macro for these. I'll add that in the next commit.

bux578 replied:
And please add him to Authors.txt while you're at it :)

MikeMatrix replied:
Will do

MikeMatrix commented:
Noticed this has still an issue with transmitting. I'll be debugging that tomorrow.

MikeMatrix commented:
At least this I will do tonight.

Glowbal commented on addons/map_gestures/functions/fnc_moduleGroupSettings.sqf:
```diff
@@ -0,0 +1,53 @@
+_leadColor = call compile ("[" + (_logic getVariable ["leadColor", ""]) + "]");
+if (isNil "_leadColor" || !((typeName _leadColor) isEqualTo "ARRAY") || {count _leadColor != 4}) exitWith {};
+_color = call compile ("[" + (_logic getVariable ["color", ""]) + "]");
+if (isNil "_color" || !((typeName _color) isEqualTo "ARRAY") || {count _color != 4}) exitWith {};
```
No validating of the content of the array.
No warning messages printed if anything was wrong.

Glowbal commented on addons/map_gestures/functions/fnc_moduleGroupSettings.sqf:
```diff
@@ -0,0 +1,53 @@
+_configurationGroupsIndex = +GVAR(GroupColorConfigurationsGroupIndex);
+
+if (isNil "_configurationGroups") then {_configurationGroups = [];};
+if (isNil "_configurationGroupsIndex") then {_configurationGroupsIndex = [];};
```
Above 4 lines could be combined, ie;
```sqf
_configurationGroups = if (isnil QGVAR(GroupColorConfigurationsGroups) then {[]}else{GVAR(GroupColorConfigurationsGroups)};
```

Glowbal commented on addons/map_gestures/functions/fnc_moduleSettings.sqf:
```diff
@@ -0,0 +1,35 @@
+[_logic, QGVAR(interval), "interval"] call EFUNC(common,readSettingFromModule);
+
+_defaultLeadColor = call compile ("[" + (_logic getVariable ["defaultLeadColor", ""]) + "]");
+if (isNil "_leadColor" || !((typeName _leadColor) isEqualTo "ARRAY") || {count _leadColor != 4}) exitWith {};
```
No validating of the content of the array.
No warning messages printed if anything was wrong.

Glowbal commented on addons/map_gestures/functions/fnc_moduleSettings.sqf:
```diff
@@ -0,0 +1,35 @@
+[QGVAR(defaultLeadColor), _defaultLeadColor, true, true] call EFUNC(common,setSetting);
+
+_defaultColor = call compile ("[" + (_logic getVariable ["defaultColor", ""]) + "]");
+if (isNil "_color" || !((typeName _color) isEqualTo "ARRAY") || {count _color != 4}) exitWith {};
```
same

Glowbal commented on addons/map_gestures/functions/fnc_receiverInit.sqf:
```diff
@@ -0,0 +1,32 @@
+{
+    if (isPlayer _x) then {
+            _nameSane = [name _x] call FUNC(sanitizeName);
+            call compile format [QUOTE(GVAR(%1_DrawPos) = [ARR_3(1,1,1)];), _nameSane];
```
looks like a double indentation here

MikeMatrix commented:
I updated the PR description, since apparently I forgot to mention that this is still a WIP code

gienkov commented:
Correct me if Im wrong, but this uses "Dot" marker right now right?
If it's **map pointing** then maybe it could be good to create simple "pointing" marker, like this one:
![pointing](http://icons.iconarchive.com/icons/iconsmind/outline/512/One-Finger-icon.png)

MikeMatrix commented:
@gienkov idk how that would improve really anything. The dots are straight forward and easy to see. I mean it wouldn't be too much effort to exchange it, I just don't see a reason to.

MikeMatrix commented:
I think this should be at least fairly stable now. Although any testing of it is appreciated.
The only thing I can think of, is getting some icons for the modules and writing better descriptions for the options.

alganthe commented:
:+1:

654wak654 commented:
What @gienkov suggested could be done with a Boolean setting, something like `isHand`, or `useHand`. The dot could be default and the pointing finger icon whould be optional.

Also shouldn't the hand icon be half the size of the image so the finger points exactly at the middle?
Demonstration:
![theory](https://cloud.githubusercontent.com/assets/8149303/9290904/7ad29be0-43ac-11e5-986d-8eaa9408c1ba.jpg)

gienkov commented:
I think it should be.

Glowbal commented on addons/map_gestures/functions/fnc_assignClientIDOnServer.sqf:
```diff
@@ -0,0 +1,24 @@
+params ["_unitName"];
+
+{
+    if (name _x == _unitName) then {
```
Can exit when you have found the unit with given name.

Glowbal commented on addons/map_gestures/functions/fnc_drawMapGestures.sqf:
```diff
@@ -0,0 +1,47 @@
+#include "script_component.hpp"
+
+if (!GVAR(enabled)) exitWith {};
+if (!visibleMap) exitWith {};
```
You can combine the above to lines into one if statement.

Glowbal commented on addons/map_gestures/functions/fnc_drawMapGestures.sqf:
```diff
@@ -0,0 +1,47 @@
+params ["_mapHandle"];
+
+_proximityPlayers = [ACE_player, GVAR(maxRange)] call FUNC(getProximityPlayers);
+{
```
Please document more with comments stating what is going on in this function.

Glowbal commented on addons/map_gestures/functions/fnc_drawMapGestures.sqf:
```diff
@@ -0,0 +1,47 @@
+        };
+    };
+    nil
+} count _proximityPlayers;
```
Uncessary variable assignment of `_proximityPlayers`. You can use `[ACE_player, GVAR(maxRange)] call FUNC(getProximityPlayers)` here directly.

Glowbal commented on addons/map_gestures/functions/fnc_getProximityPlayers.sqf:
```diff
@@ -0,0 +1,23 @@
+
+_proximityPlayers = (getPos _unit) nearEntities [["CAMAnBase"], _range];
+_proximityPlayers = _proximityPlayers - [_unit];
+(_proximityPlayers + (crew vehicle _unit))
```
Use `append` instead of binary addition.

Glowbal commented on addons/map_gestures/functions/fnc_getProximityPlayers.sqf:
```diff
@@ -0,0 +1,23 @@
+params ["_unit", "_range"];
+
+_proximityPlayers = (getPos _unit) nearEntities [["CAMAnBase"], _range];
+_proximityPlayers = _proximityPlayers - [_unit];
```
It might be better to use this instead:
```sqf
_proximityPlayers deleteAt (_proximityPlayers find _unit);
```
Though I suspect that `_unit` is always the first element on the entities array, in which case you won't need a find.

Glowbal commented on addons/map_gestures/functions/fnc_sanitizeName.sqf:
```diff
@@ -0,0 +1,30 @@
+
+_nameSanitized = [];
+{
+    if (toString [_x] in _alphabet) then {_nameSanitized pushBack _x};
```
Better might be to use `DIK_codes` to check if it's a valid character instead of a string array of the alphabet.

Also, why are other characters not valid/usable?

MikeMatrix replied:
This solution is reliant on using player names as part of variable names. That's why this is getting sanitized so harshly.

SilentSpike replied:
Couldn't you just use player UIDs?

Glowbal commented on addons/map_gestures/functions/fnc_transmit.sqf:
```diff
@@ -0,0 +1,41 @@
+{
+    _ownerID = _x getVariable QGVAR(owner_id);
+    if (isNil "_ownerID") then {
+        [0, {[_this] call FUNC(assignClientIDOnServer)}, name _x] call cba_fnc_GlobalExecute;
```
We have our own function for global execution, but better might be to switch this over to our event framework.

Glowbal commented on addons/map_gestures/functions/fnc_transmitterInit.sqf:
```diff
@@ -0,0 +1,59 @@
+GVAR(MouseMoveHandlerID) = _mapCtrl ctrlAddEventHandler ["MouseMoving", {
+    if (EGVAR(maptools,drawing_isDrawing)) exitWith {};
+    if (EGVAR(maptools,mapTool_isDragging)) exitWith {};
+    if (EGVAR(maptools,mapTool_isRotating)) exitWith {};
```
The above 3 if statements can be combined into 1.

```sqf
if (EGVAR(maptools,drawing_isDrawing) || {EGVAR(maptools,mapTool_isDragging)} || {EGVAR(maptools,mapTool_isRotating)}) exitwith {};
```

Glowbal commented on addons/map_gestures/functions/fnc_transmitterInit.sqf:
```diff
@@ -0,0 +1,59 @@
+    if (EGVAR(maptools,mapTool_isDragging)) exitWith {};
+    if (EGVAR(maptools,mapTool_isRotating)) exitWith {};
+
+    params ["_control", "_posX", "_posY"];
```
Parameters are not necessary until after the if statement on line 33

Glowbal commented on addons/map_gestures/functions/fnc_transmitterInit.sqf:
```diff
@@ -0,0 +1,59 @@
+// MouseUp EH
+if (!isNil QGVAR(MouseUpHandlerID)) then {_mapCtrl ctrlRemoveEventHandler ["MouseButtonUp", GVAR(MouseUpHandlerID)]; GVAR(MouseUpHandlerID) = nil;};
+GVAR(MouseUpHandlerID) = _mapCtrl ctrlAddEventHandler ["MouseButtonUp", {
+    params ["", "_button"];
```
Maybe also exit here if `GVAR(enabled)` is false.

Glowbal commented on addons/map_gestures/functions/fnc_receiverInit.sqf:
```diff
@@ -0,0 +1,32 @@
+        missionNamespace setVariable [format [QGVAR(%1_DrawPos), _nameSane], [1, 1, 1]];
+    };
+    nil
+} count allUnits;
```
1. Upon receiving this default value is also given. Is there a specific reason to loop through all units and assign default values upon init as well?
2. This will not work for any JIP players.

Glowbal commented on addons/map_gestures/functions/fnc_moduleGroupSettings.sqf:
```diff
@@ -0,0 +1,51 @@
+
+params ["_logic", "_units", "_activated"];
+
+diag_log "Running";
```
rouge debug print

Glowbal commented on addons/map_gestures/functions/fnc_moduleGroupSettings.sqf:
```diff
@@ -0,0 +1,51 @@
+diag_log "Running";
+
+if (!_activated) exitWith {};
+if (!isServer) exitWith {};
```
The above 2 if statements can be combined

Glowbal commented on addons/map_gestures/functions/fnc_moduleGroupSettings.sqf:
```diff
@@ -0,0 +1,51 @@
+_configurationGroupsIndex = if (isNil QGVAR(GroupColorConfigurationsGroupIndex)) then { [] } else { +GVAR(GroupColorConfigurationsGroupIndex) };
+
+_completedGroups = [];
+_configurationIndex = _configurations pushBack [_leadColor, _color];
```
Please write some comments in the code about what is going on here

Glowbal commented on addons/map_gestures/functions/fnc_drawMapGestures.sqf:
```diff
@@ -0,0 +1,47 @@
+                if (_x == leader _group) then {GVAR(defaultLeadColor)} else {GVAR(defaultColor)};
+            };
+
+            _mapHandle drawIcon ["\a3\ui_f\data\gui\cfg\Hints\icon_text\group_1_ca.paa", _color, _pos, 55, 55, 0, "", 1, 0.030, "PuristaBold", "left"];
```
On line 42 and 43 you have magic numbers (`55` and `20` for the sizes, the texture path, etc). Please put those into defines so they have a meaningful name.

Glowbal commented on addons/map_gestures/functions/fnc_transmit.sqf:
```diff
@@ -0,0 +1,41 @@
+params ["", "_pfhId"];
+
+if (!GVAR(EnableTransmit) || !visibleMap) exitWith {
+    call FUNC(endTransmit);
```
When calling `FUNC(endTransmit)` through releasing the mouse button, it will be called twice. Once from the mouse button and once while exiting the PFH.

Glowbal commented on addons/map_gestures/functions/fnc_transmit.sqf:
```diff
@@ -0,0 +1,41 @@
+    if (isNil "_ownerID") then {
+        [0, {[_this] call FUNC(assignClientIDOnServer)}, name _x] call cba_fnc_GlobalExecute;
+    } else {
+        if (_ownerID != ACE_player getVariable QGVAR(owner_id)) then {
```
Here it's being assumed that `ACE_player` always have the `owner_id` assigned. Judging from the code in this component, this may not always be the case. Remote controlled units is one example.

Glowbal commented on addons/map_gestures/ACE_Settings.hpp:
```diff
@@ -0,0 +1,50 @@
+        displayName = CSTRING(interval_displayName);
+        description = CSTRING(interval_description);
+    };
+    class GVAR(defaultLeadColor) {
```
Color settings might be better as client side settings.

MikeMatrix replied:
No, since this is part of the framework that allows mission makers to change who get's what color, based on their involvement.

Glowbal replied:
I get that. But why? Colors are imo a user specific setting. It would help color blind people to set the color to something they can see clearly. What is the benefit of having this set globaly or the reason behind it?

SilentSpike replied:
I agree, colours should be set by the client for users who have colour-specific needs.

alganthe commented:
just let the user chose the color, mission makers can still force it via settings.

MikeMatrix commented:
Alright. Fixed a few issues with settings in general and added the option to change the name text color as well as the default colors.
Being able to switch any of the other colors is just not really viable, considering those are attached to actual groups rather than anything else, so really only useful for mission makers.

Still need to test some of the changes.

jonpas commented on addons/map_gestures/ACE_Settings.hpp:
```diff
@@ -0,0 +1,74 @@
+        displayName = CSTRING(enabled_displayName);
+        description = CSTRING(enabled_description);
+        category = CSTRING(mapGestures_category);
+        isClientSettable = 0;
```
This lines can be removed, default is `0` anyways.

PabstMirror commented:
@MikeMatrix merge conflicts

MikeMatrix commented:
I'll fix it in the next few minutes.

MikeMatrix commented:
Merge conflict is resolved.
